Mongo7 upgrade

Xiangs18 opened the pull request:
This PR does the following:

- Removed all submodules (jars, kbapi_common, nms).
- Added the NMS Docker service for local testing.
- Fixed a bug in mock_auth/servicer.py.
- Updated the catalog Docker build script.
- Implemented lazy loading for MongoDB.
- Upgraded mongo and Cleaned up files in the repository.

Note: Next PR will add retryWrites and update release notes.

Xiangs18 committed:
mongo7 upgrade

Xiangs18 committed:
upgrade pymongo to 4.7.2

Xiangs18 committed:
update python to 3.9.19

Xiangs18 committed:
update Makefile and run_tests.sh scripts

Xiangs18 committed:
update submodules

Xiangs18 committed:
fix gradle not found

Xiangs18 committed:
start nms service with docker-compose

Xiangs18 committed:
update nms submodule to the latest

Xiangs18 committed:
install docker-compose in GHA

Xiangs18 committed:
switch catalog mongo7 port to 27018

Xiangs18 committed:
fix mongo port

Xiangs18 committed:
fix mongo port mapping

Xiangs18 committed:
set admin_users

Xiangs18 committed:
check admin users

Xiangs18 committed:
shut down NMS service properly

Xiangs18 committed:
fix nms curl

Xiangs18 committed:
update mock_auth service

bio-boris commented on test/mock_auth/server.py:
```diff
@@ -0,0 +1,141 @@
+    sys.exit(1)
+
+endpoints = []
+catalog_mock_auth = ['auth_admin.json', 'auth_invalid.json', 'auth_missing.json', 'auth_non_admin.json']
```
It would be good to put a comment here explaining that the order is being preserved here in order to ensure the routes capture the requests in the correct order

Xiangs18 replied:
👍

bio-boris replied:
Not sure if a comment was added or not, but additional info about "Mock server servers matches routes in a specific order, this is the current order required for tests to pass"

Xiangs18 replied:
I did add a comment. For some reason, it is not showing up here, but you can see it in the changed files.

bio-boris replied:
I see "# The order is preserved here to ensure routes capture requests correctly." this is not very descriptive and not really telling the story of why this needs to be here.

Xiangs18 replied:
Comment added. Please let me know if it makes sense now.

bio-boris replied:
LGTM!

Xiangs18 committed:
update the commit

bio-boris committed:
Update test.cfg.example

bio-boris committed:
Update test.cfg.example

bio-boris committed:
Update core_registration_test.py

Xiangs18 committed:
update message

Xiangs18 committed:
Merge branch 'dev-mongo7_upgrade' of github.com:kbase/catalog into dev-mongo7_upgrade

Xiangs18 committed:
test method_spec_admin_users

Xiangs18 committed:
test simple_good_repo old commit

Xiangs18 committed:
test register hack

Xiangs18 committed:
remove jars nms submodules

Xiangs18 committed:
update Dockerfile

Xiangs18 committed:
remove kbapi_common submodule

Xiangs18 committed:
fix deprecated mongo auth

Xiangs18 committed:
remove all submodules in catalog

Xiangs18 committed:
create MongoClient through passing in params

Xiangs18 committed:
add a comment for mock_auth server.py

Xiangs18 committed:
simplify code

Xiangs18 committed:
adjust params in MongoClient

Xiangs18 committed:
add debug log messages in list_user_favorites func

Xiangs18 committed:
try reinitiate mongoclient

Xiangs18 committed:
update __init__ in db.py file

Xiangs18 committed:
fix typo

Xiangs18 committed:
add check_db_schema

Xiangs18 committed:
add boolean flag for schema check

Xiangs18 committed:
collection direct call

Xiangs18 committed:
remove debugging message

Xiangs18 committed:
display catalog version

Xiangs18 committed:
fix failed unit test in startup_test.py

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +123,217 @@ class MongoCatalogDBI:
+        self._mongo_client_initialized = False
+        self.mongo = None
+        self.db = None
+        self.index_created = defaultdict(int)
```
Should this be a bool?

Xiangs18 replied:
I used 1 for True and 0 for False.

bio-boris replied:
Yes, so why not use a defaultdict(bool) instead?

Xiangs18 replied:
👍

Xiangs18 committed:
trigger lazy load collection

Xiangs18 committed:
clean up files

Xiangs18 commented:
This PR does the following:
1. Removed all submodules (jars, kbapi_common, nms).
2. Added the NMS Docker service for local testing.
3. Fixed a bug in mock_auth/servicer.py.
4. Updated the catalog Docker build script.
5. Implemented lazy loading for MongoDB.
6. Cleaned up files in the repository.

bio-boris commented on Makefile:
```diff
@@ -27,11 +27,6 @@ LIB_DIR = lib
 
-default: init
-
-init:
```
Does this init section need to be removed or can we keep it in?

bio-boris replied:
Or was this deprecated in favor of docker compose, and was only needed for tests?

Xiangs18 replied:
All submodules have been removed, so there is no need to pull them. If you want to keep init section, what should be included in the init?

bio-boris commented on Makefile:
```diff
@@ -102,8 +97,8 @@ setup-tests:
 	mkdir -p $(TESTLIB)/biokbase
 	mkdir -p $(TESTDIR)/nms
 	rsync -av lib/biokbase/* $(TESTLIB)/biokbase/. --exclude *.bak-*
-	rsync -av kbapi_common/lib/biokbase/* $(TESTLIB)/biokbase/.
```
If we don't need this anymore, can we delete it instead of commenting it out?

Xiangs18 replied:
removed

bio-boris commented on .github/workflows/test.yml:
```diff
@@ -21,28 +21,32 @@ jobs:
         image: mongo:${{matrix.mongo-version}}
         ports:
-            - 27017:27017
+            - 27018:27017
```
Can we keep it at 27017?

Xiangs18 replied:
port 27017 is already being used by the NMS Mongo

MrCreosote replied:
I would change the NMS port in that case. If one or the other is going to have a non standard port it should be the supporting system, not the one we're working on

bio-boris replied:
Why do we need 2 seperate mongos?

MrCreosote replied:
One presumably starts up with the docker compose to support NMS, the other starts up as part of the catalog tests. That seems good to me, the catalog shouldn't know or care about any NMS dependencies

Xiangs18 replied:
👍

bio-boris replied:
So the mongo in test.yml will be 27017 and the one in nms-compose.yaml will be 27018?

Xiangs18 replied:
NMS will be mapped to port 27018, and Catalog will be mapped into port 27017.

bio-boris replied:
Looks like this was changed to 27017

bio-boris commented on .github/workflows/test.yml:
```diff
@@ -21,28 +21,32 @@ jobs:
         with:
           python-version: ${{matrix.python-version}}
 
+      - name: Install Docker Compose
```
This is not necessary. You can use `docker compose` as it is included in docker nowadays

Xiangs18 replied:
This step is necessary.

https://github.com/kbase/catalog/actions/runs/13210034758/job/36881655854?pr=140#step:6:43

MrCreosote replied:
Try `docker compose` without the hyphen

Xiangs18 replied:
👍

bio-boris commented on test/test.cfg.example:
```diff
@@ -24,8 +24,8 @@ test-user-2 = wstester2
-# host where mongo lives, e.g. localhost:27017, for tests there should be not authentication required
-mongodb-host = localhost:27017
+# host where mongo lives, e.g. localhost:27018, for tests there should be not authentication required
+mongodb-host = localhost:27018
```
Can we keep this at 27017?

Xiangs18 replied:
port 27017 is already being used by the NMS Mongo

Xiangs18 replied:
👍

bio-boris replied:
Looks like this was changed to 27017

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +123,217 @@ class MongoCatalogDBI:
+                unique=True, sparse=False)
+
+        elif collection_name == MongoCatalogDBI._LOCAL_FUNCTIONS:
+        # Make sure we have a unique index on module_name_lc and git_commit_hash
```
It is possible to refactor this giant `elif` block using a dictionary to map collection names to their respective index creation functions. Here is an example of how you can refactor it:

```python
def _create_indexes(self, collection_name):
    """Create indexes for the given collection lazily."""
    collection = self.db[collection_name]

    index_creation_map = {
        MongoCatalogDBI._MODULE_VERSIONS: [
            ('module_name_lc', False),
            ('git_commit_hash', False),
            [('module_name_lc', ASCENDING), ('git_commit_hash', ASCENDING), True]
        ],
        MongoCatalogDBI._LOCAL_FUNCTIONS: [
            'function_id',
            [('module_name_lc', ASCENDING), ('function_id', ASCENDING), ('git_commit_hash', ASCENDING), True],
            'module_name_lc',
            'git_commit_hash'
        ],
        # Add other collections and their indexes similarly
    }

    indexes = index_creation_map.get(collection_name, [])
    for index in indexes:
        if isinstance(index, tuple):
            collection.create_index(index[0], sparse=index[1])
        elif isinstance(index, list):
            collection.create_index(index, unique=index[-1])
        else:
            collection.create_index(index)
```

This approach uses a dictionary to map collection names to a list of indexes, which reduces the need for multiple `elif` statements. Adjust the `index_creation_map` dictionary as needed to cover all your collections and their respective indexes.

MrCreosote replied:
You could potentially use an index spec to simplify as well: https://github.com/kbase/cdm-task-service/blob/main/cdmtaskservice/mongo.py#L70

Xiangs18 replied:
Done

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +123,217 @@ class MongoCatalogDBI:
+        return self._get_collection(MongoCatalogDBI._SECURE_CONFIG_PARAMS)
+
+    def _create_indexes(self, collection_name):
+        """Create indexes for the given collection lazily."""
```
It might be better to move out these into smaller functions into a file or class like db_indexes.py or something as this is an extremely long function. If possible, function's shouldn't be longer than the entire height of your screen.

Xiangs18 replied:
I moved it into db_indexes.py

bio-boris replied:
Why are the indexes created lazily instead of upon first starting the app?

bio-boris replied:
I think it is better to create them when the app starts. After that you could then remove
```
        if not self.index_created[collection_name]:
            self._create_indexes(collection_name)
            self.index_created[collection_name] = True
```
From the  _initialize_mongo_client code

Xiangs18 replied:
Because when we start the app and create collections/indexes in __init__, it would encounter the MongoClient opened before fork issue.

bio-boris replied:
What about creating them before the uswgi process is started up?

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -1274,6 +1360,12 @@ def get_secure_config_params(self, module_name):
     def check_db_schema(self):
 
+        if self._db_schema_checked:
+            return
```
This is missing coverage

Xiangs18 replied:
I moved it outside check_db_schema function.

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -1455,20 +1551,23 @@ def prepare_version_doc_for_db_2_to_3_update(self, version, module):
         # make sure we don't have any indecies on the collections
-        self.volume_mounts.drop_indexes()
-        self.client_groups.drop_indexes()
+        volume_mounts_collection = self.db[MongoCatalogDBI._VOLUME_MOUNTS]
```
I think this entire function can be deleted, as we have already upgraded the indexes?

MrCreosote replied:
AFAIK we're the only ones using the catalog, all these db version changes happened years ago, and so we can just put a note in the release notes that db version upgrades are no longer supported and delete all the upgrade methods

Xiangs18 replied:
@bio-boris, what does `entire function` refer to?

Xiangs18 replied:
Also, why do we need to update the database schema to v4? I assume the `db version` here does not refer to the MongoDB version.

bio-boris replied:
update_db_3_to_4

Xiangs18 replied:
https://github.com/kbase/catalog/issues/144

bio-boris reviewed (changes requested):
The pr description is empty, and would benefit from a copy of the changelog/release notes . The version and release notes should be updated as well

bio-boris commented on test/docker-compose_nms.yml:
```diff
@@ -0,0 +1,47 @@
+version: '3.4'
+
```
Can we re-name this file as "docker-compose?"

MrCreosote replied:
If this is a compose file for the nms rather than the catalog, calling it as such seems appropriate to me

bio-boris replied:
It is a docker compose for testing purposes. By naming it "docker-compose" it allows you to start it up without specifying the path to a file. If we had multiple docker composes, I think it makes sense to use a non standard name, but since we only have one docker-compose file, we should use the standard name.

Xiangs18 replied:
It's just for starting the NMS service for CI/local testing to run unit tests. It makes more sense to keep its name as is.

MrCreosote replied:
I'm fine with `-f <file>` in this case but I don't feel that strongly about it

bio-boris commented:
> This PR does the following:
> 
>     1. Removed all submodules (jars, kbapi_common, nms). (Why though?)
> 
>     2. Added the NMS Docker service for local testing. 
> 
>     3. Fixed a bug in mock_auth/servicer.py. (Could you add more details to this?)
> 
>     4. Updated the catalog Docker build script. (More details please)
> 
>     5. Implemented lazy loading for MongoDB. (More details please)
> 
>     6. Cleaned up files in the repository. (I don't know what this means)

This also upgrades pymongo

I think these can be added to the README.md

MrCreosote commented on .github/workflows/test.yml:
```diff
@@ -77,6 +81,10 @@ jobs:
 
+          # setup env variables in docker-compose_nms.yml file
+          echo "ADMIN_USER=$ADMIN_USER"
+          echo "ADMIN_USER=$ADMIN_USER" >> $GITHUB_ENV
```
Could you add a comment about why this env var is set up and where it's used? That should help the next person who has to work on this service understand what's going on

Xiangs18 replied:
Isn't line 84 already addressing what you're suggesting, or would you like me to add more comments?

bio-boris replied:
Is this section necessary? It seems like this variable is already being set in

```
        env:
          KBASE_CI_TOKEN: ${{ secrets.KBASE_CI_TOKEN }}
          ADMIN_USER: ${{ secrets.KBASE_BOT_USER_CI }}
```

MrCreosote replied:
> Isn't line 84 already addressing what you're suggesting,

On second look yes it is, I think I was thrown by the wording a bit. For me I think this would make it a little more clear:

" This env var is used in the catalog test docker compose file for starting NMS"

MrCreosote replied:
But maybe Boris is right and you could just set the env var in the test task rather than putting it in $GITHUB_ENV. Either way it needs explanation though

Xiangs18 replied:
@bio-boris, are you saying that I can directly access ADMIN_USER from the docker-compose_nms.yml without passing it into $GITHUB_ENV, or are you suggesting that I can simply set ADMIN_USER in the docker-compose_nms.yml?

bio-boris replied:
Yes, you can access it without passing it into the github_env as you seem to be grabbing it from the github env

Xiangs18 replied:
@MrCreosote, comment added.
@bio-boris, I'm not exactly sure how I should set it up in test.yml. Do I need a global variable? Could you please provide an example?

bio-boris replied:
I don't understand. Isn't the environment variable already there?

MrCreosote replied:
You would just add the same env stanza from install dependencies to run tests

bio-boris replied:
Suggested cleanup:

    Remove echo "ADMIN_USER=$ADMIN_USER" – This doesn't provide additional value in the script.
    Remove echo "ADMIN_USER=$ADMIN_USER" >> $GITHUB_ENV – Since ADMIN_USER is already available as an environment variable in the env: block, explicitly adding it to $GITHUB_ENV is redundant.

bio-boris replied:
Oh I understand now. Yea, you could make it "global" by putting it at the "jobs" level

Xiangs18 replied:
Hi, I removed $GITHUB_ENV, and it's working now!

bio-boris replied:
LGTM!

MrCreosote replied:
:+1:

MrCreosote commented on Dockerfile:
```diff
@@ -1,25 +1,37 @@
 COPY requirements.txt requirements.txt
-RUN source activate root && \
-    pip install -r requirements.txt
+RUN pip install -r requirements.txt
```
We should convert to pipenv / poetry / uv, but that's a later task

Xiangs18 replied:
next PR

Xiangs18 replied:
https://github.com/kbase/catalog/issues/142

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/log.py:
Is this an exact copy from kb_sdk_plus?

Xiangs18 replied:
https://kbase.slack.com/archives/C05SSJT13BN/p1738107988928679?thread_ts=1738105198.145779&cid=C05SSJT13BN

MrCreosote replied:
yes, I'm aware of that post, but it doesn't answer the question

Xiangs18 replied:
yes, from https://github.com/kbase/kb_sdk_plus/blob/main/src/java/us/kbase/templates/log.py

MrCreosote replied:
:+1:

MrCreosote commented on test/run_tests.sh:
```diff
@@ -11,16 +11,15 @@
 
 echo 'Starting Mock Auth API...'
-docker run -d --rm -v ${PWD}/mock_auth:/config -p 7777:5000 --name mock-auth mockservices/mock_json_service
+docker run -d --rm -v ${PWD}/mock_auth:/config -v ${PWD}/mock_auth/server.py:/server/server.py -p 7777:5000 --name mock-auth mockservices/mock_json_service
```
This isn't something we need to deal with now, but there are plenty of 3rd party service mocking options out there, I'm not sure why we rolled our own. Also auth2 has test mode which we could use

Xiangs18 replied:
https://github.com/kbase/catalog/issues/143

MrCreosote replied:
:+1:

MrCreosote commented on test/run_tests.sh:
```diff
@@ -11,16 +11,15 @@
 echo 'Waiting for NMS to start...'
 sleep 25
-curl -d '{"id":"1","params":[],"method":"NarrativeMethodStore.ver","version":"1.1"}' http://localhost:7125
+curl -d '{"id":"1","params":[],"method":"NarrativeMethodStore.ver","version":"1.1"}' http://localhost:7125/rpc
```
Wow these tests must not've passed for a long time...

MrCreosote commented on test/mock_auth/server.py:
This file needs some explanation at the top. Why are we replacing the file in the mock server container? Whoever comes along next to work on the catalog will need to understand the reasoning behind this

Xiangs18 replied:
Comment added.

bio-boris replied:
I don't see an explanation added here at the top as per the request that explains why we are replacing this file in the mock server container.

Xiangs18 replied:
done

MrCreosote replied:
:+1:

MrCreosote commented on test/mock_auth/server.py:
What changes have been made in this file compared to the original in the mock server container?

Xiangs18 replied:
The only change is hardcoding the catalog_mock_auth sequence instead of using os.listdir(). Comment added.

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +123,217 @@ class MongoCatalogDBI:
+        self.mongo_psswd = mongo_psswd
+        self.mongo_authMechanism = mongo_authMechanism
+
+        # MongoDB client and database initialization deferred
```
Can you add an explanation as to why this is necessary?

Xiangs18 replied:
The explanation has been added.

MrCreosote replied:
:+1:

MrCreosote replied:
I would expound on the explanation - the user may not understand why and when forking is happening

Xiangs18 replied:
This conversation is irrelevant as we are not using lazy Mongo load anymore.

MrCreosote replied:
We most certainly are. We're doing 2 mongo creations - a creation on init and a lazy creation after fork. The explanation for why we're doing that is totally relevant

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +123,217 @@ class MongoCatalogDBI:
+        self.mongo_psswd = mongo_psswd
+        self.mongo_authMechanism = mongo_authMechanism
+
+        # MongoDB client and database initialization deferred
```
How lazy is lazy? Is it not until a request is made, and therefore the service can be running in a guaranteed to fail state? If so, it seems like this class should at least test connectivity and auth at init time

MrCreosote replied:
IIUC the db class instance is created pre fork, and then the fork happens and it's no longer working. If that's the case, and if it's possible, what I would do is

* in the init, check connectivity, check db schema, create indexes, etc., and close the client.
* On the first call, create a mongo client knowing that everything is set up correctly, and set up the instance variables for the collections. Then you don't need all the property functions

MrCreosote replied:
Oh, also you probably need a thread lock around the client creation step, or multiple calls could result in multiple client creations

Xiangs18 replied:
> How lazy is lazy?

I added more comments in this [commit](https://github.com/kbase/catalog/pull/140/commits/ae1b22445f28fc5a0705f36dc2dcc6cab1e936ba). If you have any further questions, please let me know.

Could you discuss the trade-offs between your design and the current implementation?

MrCreosote replied:
The issue with the current implementation the service can start in a guaranteed to fail state. In what I propose above it can't.

Xiangs18 replied:
What is guaranteed to fail state?

MrCreosote replied:
It means that the server is up, but if anyone makes a call to it that needs to contact mongo, that call will fail

Xiangs18 replied:
Why do you think that call will fail?

MrCreosote replied:
Because the server is in a guaranteed to fail state.

Xiangs18 replied:
> On the first call, create a mongo client knowing that everything is set up correctly, and set up the instance variables for the collections. Then you don't need all the property functions

Here, without the property functions, how/where do we get the first call?

MrCreosote replied:
I'm  not sure I understand the question, but you get the first call when some process makes an HTTP request to the service.

Xiangs18 replied:
Hi @bio-boris, Gavin and I were discussing this thread: https://github.com/kbase/catalog/pull/140#discussion_r1960626709. Do you have a preference for property functions, or should we add an `ensure_mongo_client_connect function/decorator` to each API method in db.py?

Xiangs18 replied:
@MrCreosote, I used a decorator, but I'm not sure why I'm getting `pymongo.errors.InvalidOperation: Cannot use MongoClient after close.` I might try making a manual call in each function instead.

MrCreosote replied:
It sounds like your client creation function is closing the client when it's done

Xiangs18 replied:
I only manually closed it at the end of `__init__`, not during the client creation function.
https://github.com/kbase/catalog/blob/1f6024a3d6b9b084b005f72b50bf2325d1154f52/lib/biokbase/catalog/db.py#L158

MrCreosote replied:
Well, the error is telling you it's closed, so either the error is wrong or the client is getting closed

MrCreosote replied:
Or the wrong client is being used

Xiangs18 replied:
The code seems to work fine in IPython during the `__init__` step. I'm not sure why it is causing problems when I reinitialize it after the fork.

MrCreosote replied:
What about calling methods post init?

Xiangs18 replied:
https://kbase.slack.com/archives/C05SSJT13BN/p1740172424557279?thread_ts=1740165659.495869&cid=C05SSJT13BN
fixed, tested, and deployed to ci

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +123,217 @@ class MongoCatalogDBI:
+        """Initialize MongoDB client and collections lazily."""
+        if not self._mongo_client_initialized:
+            try:
+                if self.mongo_user and self.mongo_psswd:
```
Can you add a note here that this case is only tested manually?

I assume it has been tested manually, yeah?

Xiangs18 replied:
Yes, simply initiating the class will cause this test to fail now. We need to trigger a function to lazily load MongoDB. Once I make a public call, the test passes.

https://github.com/kbase/catalog/blob/377f65108eea20bbbe4a9ac37fc15301ab484585/test/startup_test.py#L38-L44

MrCreosote replied:
Did this comment go to the wrong place? It doesn't seem to have anything to do with the parent comment

Xiangs18 replied:
> Can you add a note here that this case is only tested manually?

@MrCreosote, does `this case` refer to the entire function or just lines 141-143?

MrCreosote replied:
Just the case where creds are provided

Xiangs18 replied:
Note added. It has only been manually tested.

MrCreosote replied:
:+1:

Xiangs18 committed:
remove comments from Makefile

Xiangs18 committed:
validate Install Docker Compose step in ci

Xiangs18 committed:
use docker compose

Xiangs18 committed:
map NMS to the local port 27018

Xiangs18 committed:
Add a clearer comment for the environment variable in test.yml

Xiangs18 committed:
move env to jobs level

Xiangs18 committed:
add more comments in server.py

Xiangs18 committed:
add more comments in db.py

Xiangs18 committed:
remove check_db_schema related functions

Xiangs18 committed:
revert back changes

Xiangs18 committed:
move self._db_schema_checked flag outside function

Xiangs18 committed:
refactor _create_indexes function

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,130 +125,136 @@ class MongoCatalogDBI:
+        """Create indexes for the given collection lazily."""
+        collection = self.db[collection_name]
+
+        # Get the indexes for the collection from the DBIndexes class
```
It seems that all of this code can go into the db_indexes.py code.

Xiangs18 replied:
It can. However, it makes more sense to me that in db_indexes.py, we first retrieve the indexes, and then the code to create the indexes is underneath the _create_indexes function.

Xiangs18 replied:
You mean line 242?

Xiangs18 replied:
If you mean the entire thing, it makes more sense to me to separate the steps of getting and creating indexes.

bio-boris replied:
I agree. What I am suggesting is the code that creates the indexes be encapsulated in the DBIndexes class

Xiangs18 replied:
Does it look better now?

Xiangs18 replied:
no more `elif`. Call _create_indexes function in `__init__`

bio-boris replied:
I think the original way it looked was much better. The way before elif was introduced, in which it declaratively did each function call and is less confusing then the way it is now.

Xiangs18 replied:
We are reverting to the previous method because lazy Mongo load is no longer needed. It is done.

https://github.com/kbase/catalog/blob/dev-mongo7_upgrade/lib/biokbase/catalog/db.py#L210

Xiangs18 committed:
encapsulate create_indexes in the DBIndexes class

bio-boris commented on lib/biokbase/catalog/db_indexes.py:
```diff
@@ -0,0 +1,132 @@
+from pymongo import ASCENDING
```
It seems a bit strange to create indexes based on a specific collection name being rather than just create them in one go.

Here's one way to do it
```
from pymongo import ASCENDING
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class DBIndexes:
    _INDEX_MAP = {
        "module_versions": [
            {"fields": ["module_name_lc"], "unique": False, "sparse": False},
            {"fields": ["git_commit_hash"], "unique": False, "sparse": False},
            {"fields": ["module_name_lc", "git_commit_hash"], "unique": True, "sparse": False},
        ],
        "local_functions": [
            {"fields": ["function_id"], "unique": False, "sparse": False},
            {"fields": ["module_name_lc"], "unique": False, "sparse": False},
            {"fields": ["git_commit_hash"], "unique": False, "sparse": False},
            {"fields": ["module_name_lc", "function_id", "git_commit_hash"], "unique": True, "sparse": False},
        ],
        "developers": [
            {"fields": ["kb_username"], "unique": True, "sparse": False},
        ],
        "build_logs": [
            {"fields": ["registration_id"], "unique": True, "sparse": False},
            {"fields": ["module_name_lc"], "unique": False, "sparse": False},
            {"fields": ["timestamp"], "unique": False, "sparse": False},
            {"fields": ["registration"], "unique": False, "sparse": False},
            {"fields": ["git_url"], "unique": False, "sparse": False},
            {"fields": ["current_versions.release.release_timestamp"], "unique": False, "sparse": False},
        ],
        "favorites": [
            {"fields": ["user"], "unique": False, "sparse": False},
            {"fields": ["module_name_lc"], "unique": False, "sparse": False},
            {"fields": ["id"], "unique": False, "sparse": False},
            {"fields": ["user", "id", "module_name_lc"], "unique": True, "sparse": False},
        ],
        "exec_stats_raw": [
            {"fields": ["user_id"], "unique": False, "sparse": False},
            {"fields": ["app_module_name", "app_id"], "unique": False, "sparse": True},
            {"fields": ["func_module_name", "func_name"], "unique": False, "sparse": True},
            {"fields": ["creation_time"], "unique": False, "sparse": False},
            {"fields": ["finish_time"], "unique": False, "sparse": False},
        ],
        "exec_stats_apps": [
            {"fields": ["module_name"], "unique": False, "sparse": True},
            {"fields": ["full_app_id", "type", "time_range"], "unique": True, "sparse": False},
            {"fields": ["type", "time_range"], "unique": False, "sparse": True},
        ],
        "exec_stats_users": [
            {"fields": ["user_id", "type", "time_range"], "unique": True, "sparse": False},
        ],
        "client_groups": [
            {"fields": ["module_name_lc", "function_name"], "unique": True, "sparse": False},
        ],
        "volume_mounts": [
            {"fields": ["client_group", "module_name_lc", "function_name"], "unique": True, "sparse": False},
        ],
        "secure_config_params": [
            {"fields": ["module_name_lc"], "unique": False, "sparse": False},
            {"fields": ["module_name_lc", "version", "param_name"], "unique": True, "sparse": False},
        ],
    }

    @classmethod
    def get_indexes(cls, collection_name):
        return cls._INDEX_MAP.get(collection_name, [])

    @staticmethod
    def create_indexes(collection, indexes):
        index_definitions = [
            {
                "key": [(field, ASCENDING) for field in index["fields"]],
                "unique": index.get("unique", False),
                "sparse": index.get("sparse", False),
            }
            for index in indexes
        ]
        
        collection.create_indexes(index_definitions)
        logger.info(f"Created indexes: {index_definitions}")

```

bio-boris replied:
I forget, why did this code even get changed to this instead of keeping the original code?

bio-boris replied:
https://github.com/kbase/catalog/commit/ec3064b8115f12021fc903d852d1eb41fef0ff8b Looks like the code was fine before but a bunch of else ifs were added

bio-boris replied:
It did everything in one go, rather than doing things by a specific collection.

Xiangs18 replied:
The original Mongo initialization is in __init__, and it created the collection/index in a single operation. When we switched to lazy Mongo loading, the collection/index was created only once on their first call, which is why there are multiple 'elif' statements.

Now we are switching to https://github.com/kbase/catalog/pull/140#discussion_r1947256110

bio-boris replied:
I think keeping in the "declarative" and straightforward style that it was before would make this code less confusing. In addition, this can be run as a script before the server starts up.

Xiangs18 replied:
Call _create_indexes function in __init__

bio-boris replied:
I don't understand what you are saying.

Xiangs18 replied:
There is no longer a db_indexes.py file. This thread is no longer relevant.

Xiangs18 committed:
add comment at the top of server.py file

Xiangs18 committed:
refactor db.py file

Xiangs18 committed:
add flag for mongo client

Xiangs18 committed:
update wrapper function

Xiangs18 committed:
add initialize_mongo_client function manually

Xiangs18 committed:
add _close_mongo_client function

Xiangs18 committed:
try run tests without a lock

Xiangs18 committed:
revert self.lock changes

Xiangs18 committed:
Recreate the database and collection handles after reinitializing the client

Xiangs18 committed:
revert back tests

Xiangs18 committed:
remove decorator related code

Xiangs18 committed:
add comment why need to manually close mongoclient

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +122,75 @@ class MongoCatalogDBI:
+        self.mongo_psswd = mongo_psswd
+        self.mongo_authMechanism = mongo_authMechanism
+
+        self.lock = threading.Lock()
```
I was concerned about how the fork would affect the lock object, so I asked chatgpt. It thinks that it'll be fine if it's created at the module level, so let's do that. TBH its explanations of why doing it in init is bad were pretty vague but it's probably cleaner than in the class init anyway.

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +122,75 @@ class MongoCatalogDBI:
+
+        self.lock = threading.Lock()
+        self.mongo_client = None
+        self.client_initialized = False
```
I think you can just check whether the mongo client is None or not, the extra boolean is redundant

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +122,75 @@ class MongoCatalogDBI:
+        self.client_initialized = False
+
+        # Initialize mongo client
+        self._initialize_mongo_client()
```
To be safe, I would avoid creating any mongo releated instance variables in the init function, as that could cause bugs in the future if someone doesn't deal with re-initing the client correctly. Have the init mongo method return the client and then pass it to the db schema, indexes, and close method

Xiangs18 replied:
I see you want to pass the client handle around to be safe. 
What should be returned if the client is already initialized? Should we always return a new client?

```
    def _initialize_mongo_client(self):
        """Initialize MongoDB client with lock to prevent race conditions."""
        if self.client_initialized:
            return
```

MrCreosote replied:
It can't be initialized since we're in the init method.

Xiangs18 replied:
why it can be initialized before? And you recommended that check before going to the lock.

MrCreosote replied:
It can't be initialized before, it's in the init, which is run before any other method.

The check is for post fork, when there's no need to enter the lock if the client is available.

Xiangs18 replied:
fixed, deployed, and tested in CI
https://github.com/kbase/catalog/pull/140#issuecomment-2683212105

MrCreosote replied:
:+1:

MrCreosote commented on Dockerfile:
```diff
@@ -1,25 +1,37 @@
-FROM kbase/sdkbase2:python AS build
+FROM python:3.9.19 AS build
 
+RUN apt-get update && apt-get install -y rsync
```
Can you note that the rysnc install is needed for the Makefile?

Or update the makefile to not need it, up to you

Xiangs18 replied:
added a comment

MrCreosote replied:
:+1:

MrCreosote commented on Makefile:
```diff
@@ -102,8 +97,8 @@ setup-tests:
-	cd narrative_method_store; make; make build-classpath-list;
+
+#	cd narrative_method_store; make; make build-classpath-list;
 #	rsync -av narrative_method_store/lib/biokbase/* $(TESTLIB)/biokbase/.
```
Might as well remove these comments as well

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on test/docker-compose_nms.yml:
```diff
@@ -0,0 +1,47 @@
+      - method_spec_temp_dir=narrative_method_store_temp
+      - method_spec_mongo_host=mongo:27017
+      - method_spec_mongo_dbname=method_store_repo_db
+      - method_spec_admin_users=${ADMIN_USER}
```
how do you expect a dev to deal with this when testing locally?

Xiangs18 replied:
For testing locally,  either export the variable directly in the shell, or hardcode an admin user in the YAML file.

MrCreosote replied:
How would a local dev know to do that?

Xiangs18 replied:
added a comment

MrCreosote replied:
The first time dev is still going to have a hard time figuring out why the tests don't pass when they run them. Put yourself in their shoes - if you were running the tests for the first time, what would happen if you didn't know about this env var and how easy would it be to debug? How would you want this to work / be documented?

Xiangs18 replied:
I have updated it. Let me know if this is clearer this time.

MrCreosote replied:
So again, I'm going to ask you to put yourself in the shoes of a first time developer running the tests. What will that developer see? What will they have to do to figure out what's wrong? How can we make it easier for them?

Xiangs18 replied:
Maybe place the comments at the top of the docker-compose.yml file and add a comment in the [NarrativeMethodStore] section of the test.cfg.example file as well.

If you have something else in mind, please let me know.

MrCreosote replied:
I don't have anything particular in mind because I don't understand the repercussions of not having the env var set.

Let me ask this - if a new dev runs the tests without the env var set, what will happen? What error message will they see?

Xiangs18 replied:
It will basically explain that the value set by the developer in method-spec-admin-users under the [NarrativeMethodStore] section of the test.cfg file is not a global admin, as shown below. Once a new developer sets that parameter in the docker-compose_nms.yml file, as instructed in my comment, this error will be resolved.

<img width="1293" alt="Screenshot 2025-02-26 at 4 14 12 PM" src="https://github.com/user-attachments/assets/0bddb0bf-3df2-4eaa-9ae6-2f226b25ab86" />

MrCreosote replied:
> developer in method-spec-admin-users under the [NarrativeMethodStore] section of the test.cfg

Ok, so looking at test.cfg.example, the user there is wstester4, so they'll get that message but with wstester4 instead of sijiex, right?

So how would they even know to look in the docker-compose_nms.yml file?

Xiangs18 replied:
I just ran through some experiments, and what I found is that the nms-admin-token provided by the developer in test.cfg must match what they set in the docker-compose_nms.yml file. So, essentially, when they set the nms-admin-token in test.cfg, we can add step 4) to set the corresponding method_spec_admin_users parameter in the docker-compose_nms.yml file under the # TO RUN TESTS: section. So they know!

Xiangs18 replied:
<img width="974" alt="Screenshot 2025-02-26 at 5 20 06 PM" src="https://github.com/user-attachments/assets/ae2dcd75-7515-4e67-9a66-d353bb41f8bd" />

MrCreosote replied:
That sounds reasonable. The other things I would do are:

* Get rid of the references to username / password in the TO RUN TESTS section, those no longer exist.
* Put a short section in the readme, prior to the change log, telling the user how to run tests. This is probably just something like "read the instructions at the top of test/test.cfg.example"
* Now that there's a docker compose file, is the NarrativeMethodStore section in the test cfg even used? If not it should be deleted.

Xiangs18 replied:
Do we even need step 2)? It's already set by default to `mongodb-host = localhost:27017`. Maybe we could eliminate this step as well.

MrCreosote replied:
if the tests automatically start up mongo at that url sounds like no

Xiangs18 replied:
👍

MrCreosote replied:
:

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +122,75 @@ class MongoCatalogDBI:
+        # Use the lock to ensure only one thread initializes the mongo client at a time
+        with self.lock:
+            try:
+                if not self.client_initialized or self.mongo_client is None:
```
```suggestion
        """Initialize MongoDB client with lock to prevent race conditions."""
        # Use the lock to ensure only one thread initializes the mongo client at a time
        if self.mongo_client:
            return
        with self.lock:
            try:
                if not self.client_initialized or self.mongo_client is None:
```

Don't even pay the cost of entering the lock if the client is available

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -732,15 +811,16 @@ def get_module_details(self, module_name='', git_url='', substitute_versions=Tru
 
     #### LIST / SEARCH methods
-
     def find_basic_module_info(self, query):
```
I think this blank line should stay, the above comment applies to many methods, not just this one

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -842,20 +924,22 @@ def list_all_released_service_module_versions(self):
             }))
 
     #### developer check methods
-
```
I think this blank line should stay, the above comment applies to many methods, not just this one

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

Xiangs18 committed:
update comments

Xiangs18 committed:
refactor the db.py and add comments for docker-compose file

Xiangs18 committed:
add a comment for _ensure_mongo_connection function

Xiangs18 commented:
<img width="1170" alt="Screenshot 2025-02-25 at 12 28 20 PM" src="https://github.com/user-attachments/assets/ba561841-8820-469d-9b4a-7ffacc7247f0" />

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +125,69 @@ class MongoCatalogDBI:
+
+            return mongo_client
+
+        except ConnectionFailure as e:
+            error_msg = "Cannot connect to Mongo server\n"
+            error_msg += "ERROR -- {}:\n{}".format(
+                e, "".join(traceback.format_exception(None, e, e.__traceback__))
+            )
+            raise ValueError(error_msg)
```
```suggestion
        except ConnectionFailure as e:
            raise ValueError(f"Cannot connect to Mongo server: {e}") from e
```

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +125,69 @@ class MongoCatalogDBI:
+
+    def _initialize_mongo_client(self):
+        """Initialize MongoDB client."""
+        # Use the lock to ensure only one thread initializes the mongo client at a time
```
I think this comment is in the wrong place

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -1301,34 +1416,37 @@ def check_db_schema(self):
                 'Incompatible DB versions.  Expecting DB V4, found DV V' + str(db_version) +
                 '. You are probably running an old version of the service.  Start up failed.')
 
-    def get_db_version(self):
-        # version is a collection that should only have a single 
-        version_collection = self.db[MongoCatalogDBI._DB_VERSION]
+    def get_db_version(self, db):
+
+        # version is a collection that should only have a single
+        version_collection = db[MongoCatalogDBI._DB_VERSION]
         ver = version_collection.find_one({})
         if (ver):
             return ver['version']
         else:
             # if there is no version document, then we are DB v1
-            self.update_db_version(1)
+            self.update_db_version(1, db)
             return 1
 
-    def update_db_version(self, version):
+    def update_db_version(self, version, db):
+
         # make sure we can't have two version documents
-        version_collection = self.db[MongoCatalogDBI._DB_VERSION]
+        version_collection = db[MongoCatalogDBI._DB_VERSION]
         version_collection.create_index('version_doc', unique=True, sparse=False)
         version_collection.update_one({'version_doc': True}, {'$set': {'version': version}},
                                       upsert=True)
 
     # version 1 kept released module versions in a map, version 2 updates that to a list
     # and adds dynamic service tags
-    def update_db_1_to_2(self):
-        for m in self.modules.find({'release_versions': {'$exists': True}}):
+    def update_db_1_to_2(self, db):
+        modules_collection = db[MongoCatalogDBI._MODULES]
+        for m in modules_collection.find({'release_versions': {'$exists': True}}):
             release_version_list = []
```
Why can't we delete all these upgrade methods again?

Xiangs18 replied:
https://kbase.slack.com/archives/C05SSJT13BN/p1739471119225879?thread_ts=1739292839.872669&cid=C05SSJT13BN

https://kbase.slack.com/archives/C05SSJT13BN/p1739471521100149?thread_ts=1739292839.872669&cid=C05SSJT13BN

MrCreosote replied:
Ok, so the answer is that when you remove them tests start failing. Can you make an issue in that case?

Xiangs18 replied:
https://github.com/kbase/catalog/pull/140#discussion_r1955046184

MrCreosote replied:
Oh great, forgot about that. :+1:

Xiangs18 committed:
Clarify the comments in the db.py and docker-compose.yml files

Xiangs18 committed:
remove NarrativeMethodStore section and update the test instructions in the test/test.cfg.example file

Xiangs18 committed:
expound on the explanation

MrCreosote commented:
Note to self: all of my comments are addressed at this point

MrCreosote commented on test/docker-compose_nms.yml:
```diff
@@ -0,0 +1,50 @@
+      - method_spec_git_repo_refresh_rate=2
+      - method_spec_cache_size=5000
+      - method_spec_temp_dir=narrative_method_store_temp
+      - method_spec_mongo_host=mongo:27017
```
```suggestion
      - method_spec_mongo_host=mongo:27018
```

Xiangs18 replied:
The MongoDB container still listens on port 27017 internally, and the port mapping 27018:27017 already does the job. Therefore, no changes should be made here.

MrCreosote replied:
Huh, I didn't realize until now that the connections were made internally in the containers, not via the port mapping. However, that means the database deletion below doesn't make any sense

MrCreosote replied:
:+1:

MrCreosote commented on test/docker-compose_nms.yml:
```diff
@@ -0,0 +1,50 @@
+      - service_port=7125
+    command:
+      - "-wait"
+      - "tcp://mongo:27017"
```
```suggestion
      - "tcp://mongo:27018"
```

Xiangs18 replied:
The MongoDB container still listens on port 27017 internally, and the port mapping 27018:27017 already does the job. Therefore, no changes should be made here.

MrCreosote replied:
:+1:

MrCreosote commented on test/catalog_test_util.py:
```diff
@@ -230,7 +227,9 @@ def tearDown(self):
             self.dockerclient.remove_image(image['Id'])
 
         # make sure NMS is clean after each test
-        self.mongo.drop_database(self.nms_test_cfg['method-spec-mongo-dbname'])
+        # Drop the database set by the method_spec_mongo_dbname environment variable
+        # in the docker-compose_nms.yml file.
+        self.mongo.drop_database('method_store_repo_db')
```
This is dropping a database in the catalog mongo db, not the one started up via the docker compose file. Does this run after every test or only at the end of the test suite? If the latter, I don't think it's needed since run_tests.sh does a `docker compose down`. If it is this file will need to set up a mongo client specifically for the NMS.

Xiangs18 replied:
This is indeed dropping a database in the catalog MongoDB. But why would it be placed under the [NarrativeMethodStore] section?

The tearDown class method is run at the end of each test file, so I'm not 100% sure of its intention, as it could go either way. Dropping a database in the catalog MongoDB makes more sense to me because the catalog MongoDB handle was set up in the init helper method.

If that's the case, I would suggest deleting comments L230-231 and moving method_spec_mongo_dbname under the [catalog-test] section, unless you have a different perspective.

MrCreosote replied:
> But why would it be placed under the [NarrativeMethodStore] section?

Because it's the name of the NMS database, I presume.

> Dropping a database in the catalog MongoDB makes more sense to me because the catalog MongoDB handle was set up in the init helper method.

I'm not understanding the argument here - IIUC, the original code dropped the NMS database for the NMS server. At that point, the NMS and catalog shared a Mongo instance. Now they have separate instances and so dropping the NMS database from the catalog's mongo does nothing, because there's nothing to drop. The NMS database is now on the Mongo instance started by the docker compose file, and so needs to be dropped in that Mongo instance.

Am I missing something here?

Xiangs18 replied:
> make sure NMS is clean after each test

Yeah, this comment confirms that it's actually deleting the database from the NMS.

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +125,73 @@ class MongoCatalogDBI:
-        self.mongo = MongoClient('mongodb://' + mongo_host)
+        # We're performing two MongoDB client initializations—one during the initial setup (on init), and
+        # another lazy initialization after the process is forked. This approach is necessary because MongoDB client
+        # connections are not fork-safe. When a process is forked, the child process inherits a copy of the parent’s memory,
```
```suggestion
        # connections are not fork-safe and some servers, including uwsgi, fork workers from parent processes.
        # When a process is forked, the child process inherits a copy of the parent’s memory,
```

Xiangs18 replied:
👍

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -1337,49 +1459,52 @@ def update_db_1_to_2(self):
 
             # make sure everything has the dynamic service flag
             if not 'dynamic_service' in m['info']:
-                self.modules.update_one(
+                modules_collection.update_one(
                     {'_id': m['_id']},
                     {'$set': {'info.dynamic_service': 0}})
 
             if m['current_versions']['release']:
                 if not 'dynamic_service' in m['current_versions']['release']:
-                    self.modules.update_one(
+                    modules_collection.update_one(
                         {'_id': m['_id']},
                         {'$set': {'current_versions.release.dynamic_service': 0}})
 
             if m['current_versions']['beta']:
                 if not 'dynamic_service' in m['current_versions']['beta']:
-                    self.modules.update_one(
+                    modules_collection.update_one(
                         {'_id': m['_id']},
                         {'$set': {'current_versions.beta.dynamic_service': 0}})
 
             if m['current_versions']['dev']:
                 if not 'dynamic_service' in m['current_versions']['dev']:
-                    self.modules.update_one(
+                    modules_collection.update_one(
                         {'_id': m['_id']},
                         {'$set': {'current_versions.dev.dynamic_service': 0}})
 
         # also ensure the execution stats fields have correct names
-        self.exec_stats_apps.update_many({'avg_queue_time': {'$exists': True}},
+        exec_stats_apps_colleciton = db[MongoCatalogDBI._EXEC_STATS_APPS]
+        exec_stats_apps_colleciton.update_many({'avg_queue_time': {'$exists': True}},
                                          {'$rename': {'avg_queue_time': 'total_queue_time',
                                                       'avg_exec_time': 'total_exec_time'}})
-        self.exec_stats_users.update_many({'avg_queue_time': {'$exists': True}},
+        exec_stats_apps_colleciton.update_many({'avg_queue_time': {'$exists': True}},
```
This looks like a C&P error

MrCreosote replied:
Can you run through the changes here one more time and ensure there aren't any other errors like this, where the collection is switched (assuming I'm understanding this correctly)

Xiangs18 replied:
👍

MrCreosote replied:
Does that mean you double checked all the collection changes where `self` was removed?

Xiangs18 replied:
yes, I did

MrCreosote replied:
:+1:

MrCreosote commented on lib/biokbase/catalog/db.py:
```diff
@@ -1337,49 +1459,52 @@ def update_db_1_to_2(self):
 
         # also ensure the execution stats fields have correct names
-        self.exec_stats_apps.update_many({'avg_queue_time': {'$exists': True}},
+        exec_stats_apps_colleciton = db[MongoCatalogDBI._EXEC_STATS_APPS]
+        exec_stats_apps_colleciton.update_many({'avg_queue_time': {'$exists': True}},
```
```suggestion
        exec_stats_apps_collection = db[MongoCatalogDBI._EXEC_STATS_APPS]
        exec_stats_apps_collection.update_many({'avg_queue_time': {'$exists': True}},
```

Xiangs18 replied:
Is this conversation the same as [this](https://github.com/kbase/catalog/pull/140#discussion_r1975967554)? Also, the suggested change you mentioned doesn't indicate anything different.

MrCreosote replied:
This is just a spelling correction

Xiangs18 replied:
fixed!

MrCreosote replied:
:+1:

Xiangs18 committed:
fix c&p error

Xiangs18 committed:
fix typos and clarify comments

Xiangs18 committed:
make sure NMS is clean after each test

MrCreosote reviewed (commented):
This LGTM, but @bio-boris should review / approve

bio-boris commented on Dockerfile:
```diff
@@ -1,25 +1,38 @@
 ARG VCS_REF
 ARG BRANCH
 
+RUN apt-get update && apt-get install -y wget uwsgi
```
Does `uswgi` need to be installed here? We already have it in the requirements.txt

Xiangs18 replied:
It will be addressed in the next PR.

Xiangs18 replied:
https://github.com/kbase/catalog/pull/145

bio-boris commented on lib/biokbase/catalog/db.py:
```diff
@@ -120,18 +125,75 @@ class MongoCatalogDBI:
+        self.mongo_db = mongo_db
+        self.mongo_user = mongo_user
+        self.mongo_psswd = mongo_psswd
+        self.mongo_authMechanism = mongo_authMechanism
```
This should be _ case and not camelCase in order to maintain consistency

Xiangs18 replied:
It will be addressed in the next PR.

bio-boris reviewed (approved):
Looks good enough to me to merge. We should remove uswgi from the dockerfile in the next PR i think
